### DISCUSSION TITLE
Multiple optimizable design regions in meep adjoint

### DISCUSSION THREAD

**Discussion by yugeniom** [NanoComp/meep#2284] [Q&A]
Hello,
I would like to implement a reverse engineered optimization, implementing disjoint design regions or adiacent regions with different material grids (say I want to optimize the depth of a shallow etched grating across a multilayer structure made of different materials, as an example); is this possible within the current implementation of meep adjoint? I noticed in this example script (https://nbviewer.org/github/NanoComp/meep/blob/master/python/examples/adjoint_optimization/01-Introduction.ipynb) that the design region is passed to the Optimization object under the name "design_regionS" with a plural, as if it was possible to implement something like multiple optimizable regions.

**Comment by smartalecH** [comment]
Sure, you can have as many design regions as you want. See [our Meep adjoint paper](https://doi.org/10.1364/OE.442074), along with a [recent grating coupler paper](https://doi.org/10.1364/OE.466015). 

You pass a list of all your design regions to the `OptimizationProblem` object. When you call the object itself, passing the design parameter (eg `opt(x)`), you pass a list of arrays, where each array maps to a design region. 

In your case it may be easier to just do one big 3D design region, and do a 3D parameter fit. But it's worth trying it out with multiple design regions.

**Reply by yugeniom** [reply]
Thank you very much!
A corollary to my question: does meep adjoint only allow topology optimization on a dielectric volume or would it be possible to get the AD gradients for custom parametrizations? Namely I would like to sweep the etching depth of a grating, but this would be an hyperparameter in a topology optimization setting;  could it be possible to discrete-parametrize the problem as the etching depth variable plus one variable for each width of  teeth and trenches of the grating, while retaining an AD retrieval of the gradients? Or should I implement it turning to finite difference retrieval of the gradients?

**Reply by smartalecH** [reply]
So it looks like you are really asking two questions:

**Question 1:** Can we compute gradients for density-based **hyperparameters**, like the design region size itself?

Not with the current framework, unfortunately. This is why I suggested you instead do a **3D optimization** over a large volume (e.g. covering the maximum expected change in design region size).

**Question 2:** Can we abstract away the density-based parameterization, and instead optimize over over parameters that are more relevant to my problem (e.g. grating pitch, etch depth, slant angle, etc.)

Keep in mind that Meep's adjoint framework was originally setup explicitly for density-based topology optimization (where every pixel is a degree of freedom). So conventionally, we would say this sort of abstraction is not possible, because it usually requires a *level-set* representation. *However*... our `MaterialGrid` structure now supports subpixel smoothing, which indirectly turned it into a level-set function! Which means you _can_ (theoretically) map your parameters of interest to the individual design variables (and backprop accordingly).

Keep in mind this feature is still pretty raw with unexpected bugs. Furthermore, we don't (currently) have any tutorials that do exactly this (so you'll be wandering in the dark on your own here...) but you can check out [this comment](https://github.com/NanoComp/meep/discussions/2002#discussioncomment-2361602), [this discussion](https://github.com/NanoComp/meep/discussions/2067), and [this issue](https://github.com/NanoComp/meep/issues/2235) for some more info regarding this feature.

**Reply by yugeniom** [reply]
Hello, thanks again for the prompt answer!
I am probably putting a third different question here: in the meanwhile me or - more  probably - someone else succeeds in implementing a differentiable levelset remapping , I was thinking about gradient-free optimizing a low-parametrized grating by guiding meep thorugh a simple evolutive algorithm, then using the result to seed the meep-native topology-based optimization; I  have a doubt about a 3d optimiztion  as you propose (your answer to "Question 1"), because for thin and well contrasted layers, an effective index approximation would be not so good maybe for me; instead I was considering the idea of constraining the refractive index variables to a common "normalized breadboard": say I have three different thin layers the etching runs through; my refractive index "breadboards" (i.e. the MaterialGrids somehow) are swept in the ranges [1,na], [1,nb], [1,nc], each breadborad with the same dimensionalty (Nx x Ny); basically I would like to have the topology optimization to actually otpimize only on one "normalized" etching breadboard [0,1], rempping to the respective single-layers refractive index ranges; in other words, despite having in thks example three layers and three Design regions, I would like the topology-optimization to only optimize on (Nx x Ny) variables (the digital etching pattern) instead  of 3 x (Nx x Ny) variables; do you thing something like this would be possible and simpler to implement in current framework then the levelset remapping? Of course, I understand that I would drop the possibility to include the etching depth like this, but maybe it could be enough to have it pre-optimized by the previous gradient-free stage...

**Reply by smartalecH** [reply]
> I was thinking about gradient-free optimizing a low-parametrized grating

I would open a separate discussion regarding this. `MaterialGrid`s are meant for large-scale, **gradient-based** optimization. You're better off using the other geometric data structures for something like this.

> I have a doubt about a 3d optimiztion as you propose (your answer to "Question 1"), because for thin and well contrasted layers, an effective index approximation would be not so good maybe for me;

I'm not sure I follow... I wasn't suggesting any effective index approximations. Quite the contrary. Do a full 3D simulation (which is required for 3D optimization). My comment was in contrast to what we typically do with integrated photonics TO: 2D optimizations (since we have planar layers) but still requiring 3D simulations. The advantage here is now you aren't restricted by the design region thickness. If you have a huge volume, you can explicitly optimize over the thickness of the structure.

> considering the idea of constraining the refractive index variables to a common "normalized breadboard"

People have done this with meep in the past (see [Apratim's meepcon talk ](https://www.youtube.com/watch?v=VeCl5THchAA)and [Rajesh's paper](https://doi.org/10.1038/nphoton.2015.80)).

As an aside, you typically want to use gradient-based methods as the number of design variables increases.
